paddle/phi/kernels/impl/einsum_impl.h

```diff
@@ -0,0 +1,582 @@
+#include "paddle/utils/string/string_helper.h"
+
+namespace phi {
+// check the validation of the Einsum equation.
+// 1. the label must between 'a' - 'z'.
+// 2. the dim of the same label must be same.
+// 3. the broad cast dims in two operands is broadcastable.
+// 4. there must exist '->' and the default output is complete in python.
+// may be we can skip validation check in C++ and just put it in python.
+inline static void ValidationCheck(const std::string& equation) {
+  auto n_part = paddle::string::split_string(equation, "->").size();
+  PADDLE_ENFORCE_EQ(
+      n_part,
+      2,
+      phi::errors::InvalidArgument("must have one `->` in equation."));
+  size_t pos;
+  auto trimed_equ = equation;
+  if ((pos = trimed_equ.find("->", 0)) != std::string::npos) {
+    trimed_equ.replace(pos, 2, ".");
+  }
+  auto is_valid_char = [](char c) {
+    if (c >= 'a' && c <= 'z') return true;
+    if (c == '.' || c == ',') return true;
+    return false;
+  };
+  for (auto c : trimed_equ) {
+    if (!is_valid_char(c))
+      PADDLE_THROW(phi::errors::InvalidArgument(
+          "Invalide char in equation. einsum only accept `a`-`z` and `...`"
+          "but get:`%c`",
+          c));
+  }
+}
+
+enum LabelType {
+  ALL_TYPE = 0,
+  Batch = 1,    // ABO
+  Free,         // AO, BO
+  Contraction,  // AB
+  Reduction,    // A, B
+};
+
+// map a label('a' - 'z') -> int, O(1) speed.
+class LabelMap {
+  constexpr static int N =
+      26 + 1;  // 'a' - 'z' + '.', '.' is for broadcast dims
+  int default_value;
+  int map[N];
+
+ public:
+  explicit LabelMap(int default_value = 0) {
+    this->default_value = default_value;
+    for (int i = 0; i < N; ++i) map[i] = default_value;
+  }
+  int& operator[](int label) {
+    int i = label - 'a';
+    if (label == '.') i = N - 1;
+    return map[i];
+  }
+  int operator[](int label) const {
+    int i = label - 'a';
+    if (label == '.') i = N - 1;
+    return map[i];
+  }
+  // non-exist is present by is_default
+  bool is_default(char label) {
+    return (*this)[static_cast<int>(label)] == default_value;
+  }
+};
+
+inline void print_label(const std::vector<char>& all_labels,
+                        const LabelMap& label2type) {
+  std::string str;
+  for (int a : all_labels) {
+    std::stringstream ss;
+    ss << label2type[a];
+    str += ss.str();
+  }
+  VLOG(5) << "Label: " + str;
+}
+
+inline static void ReplaceEllipsis(std::string& s) {  // NOLINT
+  size_t pos;
+  if ((pos = s.find("...", 0)) != std::string::npos) {
+    s.replace(pos, 3, ".");
+  }
+  // remove all the space in the expression
+  while ((pos = s.find(" ", 0)) != std::string::npos) {
+    s.replace(pos, 1, "");
+  }
+}
+
+inline std::vector<char> union_labels(const std::vector<char>& a,
+                                      const std::vector<char>& b) {
+  LabelMap counter(0);
+  std::vector<char> res;
+  auto f = [&](char c) {
+    if (counter[static_cast<int>(c)] == 0) res.push_back(c);
```

Comment:
按照内部规范，如果if block 不是最后一行，则需要用{}
```suggestion
    if (counter[static_cast<int>(c)] == 0) { 
        res.push_back(c);
     }
```